### DISCUSSION TITLE
Generate properties on one interface/model from a model's properties?

### DISCUSSION THREAD

**Discussion by Garbee** [microsoft/typespec#2953] [Q&A]
# Question

Is it possible to generate keys to one definition from the keys of another?

# Example

Given the following definition:

```typespec
@route("/user-management")
namespace UserManagement {
  model User {
    @minLength(1)
    @maxLength(255)
    firstName: string;

    @minLength(1)
    @maxLength(255)
    middleName?: string;

    @minLength(1)
    @maxLength(255)
    lastName: string;

    @minLength(4)
    @maxLength(254)
    username: string;
  }

  @get
  op getUser(@path userId: string): User;

  @post
  op updateUser(@path userId: string, @body user: User): {
    @statusCode statusCode: 200;
  } | {
    @statusCode statusCode: 400;
    @body formErrors: {

    }
  } | {
    @statusCode statusCode: 404;
  };
};
```

The goal is to make an object for `@body formErrors` on a 400 status code. This means validation of the data submitted has failed. We would like to have is something that says, "For any properties in the User model, they may be present here with a value that is a string or array of strings." Then we may also add extra manual properties like `reason: "Validation Failed" to supplement the object.

I can't find any clear way of doing this from the docs available. Is this kind of thing possible with the currently available definitions in TypeSpec? I know I could do it using Typescript with something like the following:

```typescript
type PropertyErrors<T> = {
  reason: "Validation Failed";
  [P in keyof T]?: string | string[];
};
```

**Comment by bterlson** [comment]
Unfortunately we don't yet have support for mapped types, so there is no way to compute this type transformation in TypeSpec itself. You can write a decorator that affects this transformation, but it is fairly annoying. You can look at how e.g. the [withVisibility](https://github.com/microsoft/typespec/blob/main/packages/compiler/src/lib/decorators.ts#L852) decorator works. But combined with a template, you could use this to build the template you want.
